### PR TITLE
+ extras: DSL for navigating between activities #62

### PR DESCRIPTION
**Pull request by rickrickrickyyy**
DSL for goto an activity 

``` scala
goto(classOf[SomeActivity], "SOME_ID" -> ID)
goto(classOf[SomeActivity], i=>i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP),"SOME_ID" -> ID)

### CODE REVIEW

**Comment by javipacheco** [comment]
Personally, I don't like using `Parcelable` for that. I think that we should use case classes 

```scala
case class Strings(array: String*)
```

Then you can use

```scala
goto(classOf[SomeActivity], "array-key" -> Strings("foo", "bar"))
```
Then you can use the properly method `putExtra(String name, String[] value)` for arrays in your [Intent](https://developer.android.com/reference/android/content/Intent.html), obviously you have to add the cases in the match

@fedefernandez thoughts?

File: macroid-extras/src/main/scala/macroid/extras/UIActionsExtras.scala [hunk 1]
```diff
@@ -40,6 +41,261 @@ object UIActionsExtras {
...
+  /**
+    * example: goto(classOf[TuningActivity], i ⇒  i.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP), "ID" -> 1)
+    *
+    * @param cls      class of the activity
+    * @param configFn extra config steps(add flags)
+    * @param extra    the extra key value pair
+    * @param context  context
+    * @return Ui action that can navigate to the class of the Activity
+    */
+  def goto(cls: Class[_ <: Activity], configFn: Intent ⇒  Unit, extra: (String, Any)*)(implicit context: ContextWrapper): Ui[Unit] = {
```
Comment (fedefernandez) [inline macroid-extras/src/main/scala/macroid/extras/UIActionsExtras.scala]: IMHO this `configFn` should be `Intent ⇒  Intent` and then:
```scala
Ui(context.getOriginal.startActivity(configFn(intent)))
```

File: macroid-extras/src/main/scala/macroid/extras/UIActionsExtras.scala [hunk 2]
```diff
@@ -40,6 +41,261 @@ object UIActionsExtras {
...
+  /**
+    * example: goto(classOf[TuningActivity], i ⇒  i.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP), "ID" -> 1)
+    *
+    * @param cls     class of the activity
+    * @param extra   the extra key value pair
+    * @param context context
+    * @return Ui action that can navigate to the class of the Activity
+    */
+  def goto(cls: Class[_ <: Activity], extra: (String, Any)*)(implicit context: ContextWrapper): Ui[Unit] = {
+    val intent = new Intent(context.getOriginal, cls)
```
Comment (fedefernandez) [inline macroid-extras/src/main/scala/macroid/extras/UIActionsExtras.scala]: We could call to the overloaded function
```scala
goto(cls, _ => _, extra)
```

File: macroid-extras/src/main/scala/macroid/extras/UIActionsExtras.scala [hunk 3]
```diff
@@ -40,6 +41,261 @@ object UIActionsExtras {
...
+    */
+  def goto(cls: Class[_ <: Activity], configFn: Intent ⇒  Unit, extra: (String, Any)*)(implicit context: ContextWrapper): Ui[Unit] = {
+    val intent = new Intent(context.getOriginal, cls)
+    extra.foreach(elem ⇒  elem._2.putExtra(intent, elem._1))
+    configFn(intent)
+    Ui(context.getOriginal.startActivity(intent))
+  }
+
+  /**
+    * example: goto(classOf[TuningActivity], i ⇒  i.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP), "ID" -> 1)
```
Comment (fedefernandez) [inline macroid-extras/src/main/scala/macroid/extras/UIActionsExtras.scala]: Please, remove the function argument

**Comment by rickrickrickyyy** [comment]
when trying to get the extra in the target activity using parcelable is less code

``` scala
   //  using case class
    val serializable = getIntent.getSerializableExtra("ARRAY_KEY")
    val strings = serializable.asInstanceOf[Strings]
    //using parcelable
    val strings = getIntent.getParcelableExtra[Strings]("ARRAY_KEY")
```
And because case class was serializable and serializable was already covered in the pattern matching, therefore we don't need to add the cases in the match. And we can always create the case class in the client code, therefore I think we don't need to provide the case class or parcelable class. 



I have also created another version that will call to the overloaded function base on @fedefernandez 's feedback.

``` scala
   def intentFor(cls: Class[_ <: Activity], extra: (String, Any)*)(implicit context: ContextWrapper): Intent = {
    val intent = new Intent(context.getOriginal, cls)
    extra.foreach({
      case (key, value) ⇒ value.pushInto(intent, key)
    })
    intent
  }

  def goto(intent: Intent)(implicit context: ContextWrapper): Ui[Unit] = {
    Ui(context.getOriginal.startActivity(intent))
  }

  def goto(cls: Class[_ <: Activity], extra: (String, Any)*)(implicit context: ContextWrapper): Ui[Unit] = {
    goto(intentFor(cls, extra: _*))
  }

  def goto(cls: Class[_ <: Activity], modified: Intent => Intent, extra: (String, Any)*)(implicit context: ContextWrapper): Ui[Unit] = {
    goto(modified(intentFor(cls, extra: _*)))
  }

  def request(requestCode: Int, intent: Intent)(implicit context: ActivityContextWrapper): Ui[Unit] = {
    Ui(context.getOriginal.startActivityForResult(intent, requestCode))
  }

  def request(requestCode: Int, cls: Class[_ <: Activity], extra: (String, Any)*)(implicit context: ActivityContextWrapper): Ui[Unit] = {
    request(requestCode, intentFor(cls, extra: _*))
  }

  def request(requestCode: Int, modified: Intent => Intent, cls: Class[_ <: Activity], extra: (String, Any)*)(implicit context: ActivityContextWrapper): Ui[Unit] = {
    request(requestCode, modified(intentFor(cls, extra: _*)))
  }
```

Could you please help me review this code and give me more feedback?
And also could you tell me how to create another "PR" that will change this PR?

**Comment by javipacheco** [comment]
Hi,

We have been reviewing your code, we like your proposal, but we want to give you a new approach in order to be more consistent with the rest of the code on Macroid. 

First, we SHOULDN'T use parcelable or serializabe for arrays. The Intent API has methods for arrays, and we should use these methods

```java
Intent putExtra(String name, int[] value)
Intent putExtra(String name, String[] value)
Intent putExtra(String name, short[] value)
```

We think that the new case class is a good approach for pattern matching, but if you have another propose, we can discuss it

Secondly, we think that another approach using something like a `builder` is good. For example,

```scala
goto(classOf[MyActivity])
  .withExtras("id" -> 2, "reload" -> true)
  .withFlags(FLAG_ACTIVITY_SINGLE_TOP)
  .start()
```

What's it your opinion about that?

**Comment by rickrickrickyyy** [comment]
Hi javipacheco
You were right, we SHOULDN'T use parcelable or serializabe for arrays.  I have trusted IntelliJ too much.
And as for the builder, Intent class already implement that pattern .So IMHO the implementation below will be more cohesive with the system. 

example: 
``` scala
    classOf[MainActivity]
      .withExtra("id" -> someId)
      .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
      .startActivity()
```
implementation:
```scala
  private def intentFor(cls: Class[_], extra: (String, Any)*)(implicit context: ContextWrapper): Intent = 
  {
    val intent = new Intent(context.getOriginal, cls)
    extra.foreach({
      case (key, value) ⇒ value.pushInto(intent, key)
    })
    intent
  }

 // could add more method later 
  implicit class serviceClassMethod(cls: Class[_ <: Service]) {
    def withExtra(extras: (String, Any)*)(implicit context: macroid.ContextWrapper) = {
      intentFor(cls, extras: _*)
    }

    def startService()(implicit context: macroid.ContextWrapper): Ui[Unit] = {
      intentFor(cls).startService()
    }
  }

  implicit class activityClassMethod(cls: Class[_ <: Activity]) {
    def withExtra(extras: (String, Any)*)(implicit context: macroid.ContextWrapper) = {
      intentFor(cls, extras: _*)
    }

    def startActivity()(implicit context: macroid.ContextWrapper): Ui[Unit] = {
      intentFor(cls).startActivity()
    }
  }

  implicit class intentMethod(intent: Intent) {
    def startActivity()(implicit context: macroid.ContextWrapper): Ui[Unit] = {
      Ui(context.getOriginal.startActivity(intent))
    }
    def startService()(implicit context: macroid.ContextWrapper): Ui[Unit] = {
      Ui(context.getOriginal.startService(intent))
    }
  }
```

**Comment by rickrickrickyyy** [comment]
I just don't like the goto statement with the builder approach.
```scala
     def intentOf[A <: Activity]: Class[A] = classOf[A]

     intentOf[MainActivity].start()
```

File: macroid-extras/src/main/scala/macroid/extras/UIActionsExtras.scala [hunk 4]
```diff
@@ -40,6 +41,74 @@ object UIActionsExtras {
     }
   }
 
+  private implicit class IntentVisitor(intent: Intent) {
```
Comment (rickrickrickyyy) [inline macroid-extras/src/main/scala/macroid/extras/UIActionsExtras.scala]: Once this is final,I will try to add comment in it.

**Pull request closed without merging**